client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c

```diff
@@ -0,0 +1,233 @@
+
+
+
+struct ext_log_storage_t {
+    kaa_list_t     *logs;           /**< List of @link ext_log_record_t @endlink */
+    size_t          occupied_size;  /**< Currently occupied logs volume */
+    kaa_logger_t   *logger;         /**< Logger instance */
+};
+
+
+
+void log_record_destroy(void *record_p)
+{
+    if (record_p) {
+        KAA_FREE(((ext_log_record_t*)record_p)->data);
+        KAA_FREE(record_p);
+    }
+}
+
+
+
+kaa_error_t ext_log_storage_create(ext_log_storage_t** log_storage_p, kaa_logger_t *logger)
+{
+    KAA_RETURN_IF_NIL(log_storage_p, KAA_ERR_BADPARAM);
+
+    *log_storage_p = (ext_log_storage_t *) KAA_MALLOC(sizeof(ext_log_storage_t));
+    KAA_RETURN_IF_NIL(*log_storage_p, KAA_ERR_NOMEM);
+
+    (*log_storage_p)->logs = NULL;
+    (*log_storage_p)->occupied_size = 0;
+    (*log_storage_p)->logger = logger;
+
+    return KAA_ERR_NONE;
+}
+
+
+
+void ext_log_storage_destroy(ext_log_storage_t *self)
+{
+    if (self) {
+        kaa_list_destroy(self->logs, &log_record_destroy);
+        KAA_FREE(self);
+    }
+}
+
+
+
+kaa_error_t ext_log_storage_allocate_log_record_buffer(ext_log_storage_t *self, kaa_log_record_t *record)
+{
+    KAA_RETURN_IF_NIL2(record, record->size, KAA_ERR_BADPARAM);
+
+    record->data = (char *) KAA_MALLOC(record->size * sizeof(char));
+    return KAA_ERR_NONE;
+}
+
+
+
+kaa_error_t ext_log_storage_add_log_record(ext_log_storage_t *self, kaa_log_record_t *record)
+{
+    KAA_RETURN_IF_NIL2(self, record, KAA_ERR_BADPARAM);
+
+    ext_log_record_t *new_record = (ext_log_record_t *) KAA_MALLOC(sizeof(ext_log_record_t));
+    KAA_RETURN_IF_NIL(new_record, KAA_ERR_NOMEM);
+
+    new_record->data = record->data;
+    new_record->size = record->size;
+    new_record->bucket_id = 0;
+
+    kaa_list_t *record_position = self->logs
+            ? kaa_list_push_back(self->logs, new_record)
+            : (self->logs = kaa_list_create(new_record));
+    if (!record_position) {
+        KAA_FREE(new_record);
+        return KAA_ERR_NOMEM;
+    }
+
+    record->data = NULL;
+    record->size = 0;
+    self->occupied_size += new_record->size;
+    return KAA_ERR_NONE;
+}
+
+
+
+kaa_error_t ext_log_storage_deallocate_log_record_buffer(ext_log_storage_t *self, kaa_log_record_t *record)
+{
+    KAA_RETURN_IF_NIL2(record, record->data, KAA_ERR_BADPARAM);
+
+    KAA_FREE(record->data);
+    record->data = NULL;
+    return KAA_ERR_NONE;
+}
+
+
+
+bool logs_list_find_by_bucket_id(void *log_record_p, void *bucket_id_p)
+{
+    return ((ext_log_record_t *)log_record_p)->bucket_id == *((uint16_t *)bucket_id_p);
+}
+
+
+
+kaa_error_t ext_log_storage_write_next_record(ext_log_storage_t *self, char *buffer, size_t buffer_len
+        , uint16_t bucket_id, size_t *record_len)
+{
+    KAA_RETURN_IF_NIL5(self, buffer, buffer_len, bucket_id, record_len, KAA_ERR_BADPARAM);
+
+    uint16_t zero_bucket_id = 0;
+    kaa_list_t *record_position = kaa_list_find_next(self->logs, &logs_list_find_by_bucket_id, &zero_bucket_id);
```

Comment:
To reduce overhead I propose to remember the position of the unmarked record and use it (if not null) on the next call.